I have tested the following commands to upload to testpypi, and both work with Python 3.11. and Python 3.12.

```
twine --no-color upload dist/* --repository testpypi
twine upload dist/* --repository testpypi
```

```bash
root@eca5ed1e5006:/src# twine --no-color upload dist/* --repository testpypi
Uploading distributions to https://test.pypi.org/legacy/
WARNING  Error getting password from keyring
         Traceback (most recent call last):
           File "/usr/local/lib/python3.12/site-packages/twine/auth.py", line 81, in get_password_from_keyring
             return cast(str, keyring.get_password(system, username))
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           File "/usr/local/lib/python3.12/site-packages/keyring/core.py", line 56, in get_password
             return get_keyring().get_password(service_name, username)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           File "/usr/local/lib/python3.12/site-packages/keyring/backends/fail.py", line 28, in get_password
             raise NoKeyringError(msg)
         keyring.errors.NoKeyringError: No recommended backend was available. Install a recommended 3rd party backend package; or, install the
         keyrings.alt package if you want to use the non-recommended backends. See https://pypi.org/project/keyring for details.
Enter your API token:
Uploading clang_tools-0.11.1.post1.dev5-py3-none-any.whl
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.9/19.9 kB • 00:00 • 19.3 MB/s
Uploading clang_tools-0.11.1.post1.dev5.tar.gz
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.1/79.1 kB • 00:00 • 171.3 MB/s

View at:
https://test.pypi.org/project/clang-tools/0.11.1.post1.dev5/
root@eca5ed1e5006:/src# python3 -V
Python 3.12.3
root@eca5ed1e5006:/src#
```

I'll try again soon to find the cause.

--- a/twine-upload.sh
+++ b/twine-upload.sh
@@ -120,7 +120,7 @@ if [[ ${INPUT_VERIFY_METADATA,,} != "false" ]] ; then
     twine check ${INPUT_PACKAGES_DIR%%/}/*
 fi
 
-TWINE_EXTRA_ARGS="--disable-progress-bar"
+TWINE_EXTRA_ARGS="--disable-progress-bar $TWINE_EXTRA_ARGS"
 if [[ ${INPUT_SKIP_EXISTING,,} != "false" ]] ; then
     TWINE_EXTRA_ARGS=--skip-existing
 fi

--- a/twine-upload.sh
+++ b/twine-upload.sh
@@ -39,6 +39,7 @@ INPUT_PACKAGES_DIR="$(get-normalized-input 'packages-dir')"
 INPUT_VERIFY_METADATA="$(get-normalized-input 'verify-metadata')"
 INPUT_SKIP_EXISTING="$(get-normalized-input 'skip-existing')"
 INPUT_PRINT_HASH="$(get-normalized-input 'print-hash')"
+INPUT_DISABLE_PROGRESS_BAR="$(get-normalized-input 'disable-progress-bar')"
 
 PASSWORD_DEPRECATION_NUDGE="::error title=Password-based uploads deprecated::\
 Starting in 2024, PyPI will require all users to enable Two-Factor \
@@ -132,6 +133,8 @@ fi
 if [[ ${INPUT_PRINT_HASH,,} != "false" || ${INPUT_VERBOSE,,} != "false" ]] ; then
     python /app/print-hash.py ${INPUT_PACKAGES_DIR%%/}
 fi
+
+TWINE_EXTRA_ARGS="--disable-progress-bar $TWINE_EXTRA_ARGS"
 
 TWINE_USERNAME="$INPUT_USER" \
 TWINE_PASSWORD="$INPUT_PASSWORD" \

--- a/twine-upload.sh
+++ b/twine-upload.sh
@@ -39,7 +39,6 @@ INPUT_PACKAGES_DIR="$(get-normalized-input 'packages-dir')"
 INPUT_VERIFY_METADATA="$(get-normalized-input 'verify-metadata')"
 INPUT_SKIP_EXISTING="$(get-normalized-input 'skip-existing')"
 INPUT_PRINT_HASH="$(get-normalized-input 'print-hash')"
-INPUT_DISABLE_PROGRESS_BAR="$(get-normalized-input 'disable-progress-bar')"
 
 PASSWORD_DEPRECATION_NUDGE="::error title=Password-based uploads deprecated::\
 Starting in 2024, PyPI will require all users to enable Two-Factor \
@@ -121,7 +120,7 @@ if [[ ${INPUT_VERIFY_METADATA,,} != "false" ]] ; then
     twine check ${INPUT_PACKAGES_DIR%%/}/*
 fi
 
-TWINE_EXTRA_ARGS=
+TWINE_EXTRA_ARGS="--disable-progress-bar"
 if [[ ${INPUT_SKIP_EXISTING,,} != "false" ]] ; then
     TWINE_EXTRA_ARGS=--skip-existing
 fi
@@ -133,8 +132,6 @@ fi
 if [[ ${INPUT_PRINT_HASH,,} != "false" || ${INPUT_VERBOSE,,} != "false" ]] ; then
     python /app/print-hash.py ${INPUT_PACKAGES_DIR%%/}
 fi
-
-TWINE_EXTRA_ARGS="--disable-progress-bar $TWINE_EXTRA_ARGS"
 
 TWINE_USERNAME="$INPUT_USER" \
 TWINE_PASSWORD="$INPUT_PASSWORD" \

